dpnp/dpnp_iface_mathematical.py

```diff
@@ -1682,51 +1845,206 @@ def fmod(x1, x2, /, out=None, *, where=True, dtype=None, subok=True, **kwargs):
+            [1. , 1. , 1. ]]))
+
+    In this example the spacing is also specified:
+    uniform for axis=0 and non uniform for axis=1
+
+    >>> dx = 2.
+    >>> y = np.array([1., 1.5, 3.5])
+    >>> np.gradient(np.array([[1, 2, 6], [3, 4, 5]], dtype=float), dx, y)
+    (array([[ 1. ,  1. , -0.5],
+            [ 1. ,  1. , -0.5]]),
+     array([[2. , 2. , 2. ],
+            [2. , 1.7, 0.5]]))
+
+    It is possible to specify how boundaries are treated using `edge_order`
+
+    >>> x = np.array([0, 1, 2, 3, 4])
+    >>> f = x**2
+    >>> np.gradient(f, edge_order=1)
+    array([1., 2., 4., 6., 7.])
+    >>> np.gradient(f, edge_order=2)
+    array([0., 2., 4., 6., 8.])
+
+    The `axis` keyword can be used to specify a subset of axes of which the
+    gradient is calculated
+
+    >>> np.gradient(np.array([[1, 2, 6], [3, 4, 5]], dtype=float), axis=0)
+    array([[ 2.,  2., -1.],
+           [ 2.,  2., -1.]])
 
     """
 
-    x1_desc = dpnp.get_dpnp_descriptor(x1, copy_when_nondefault_queue=False)
-    if x1_desc and not kwargs:
-        if len(varargs) > 1:
-            pass
-        elif len(varargs) == 1 and not isinstance(varargs[0], int):
-            pass
+    dpnp.check_supported_arrays_type(f)
+    ndim = f.ndim  # number of dimensions
+
+    if axis is None:
+        axes = tuple(range(ndim))
+    else:
+        axes = normalize_axis_tuple(axis, ndim)
+
+    dx = _gradient_build_dx(f, axes, *varargs)
+    if edge_order > 2:
+        raise ValueError("'edge_order' greater than 2 not supported")
+
+    # Use central differences on interior and one-sided differences on the
+    # endpoints. This preserves second order-accuracy over the full domain.
+    outvals = []
+
+    # create slice objects --- initially all are [:, :, ..., :]
+    slice1 = [slice(None)] * ndim
+    slice2 = [slice(None)] * ndim
+    slice3 = [slice(None)] * ndim
+    slice4 = [slice(None)] * ndim
+
+    otype = f.dtype
+    if dpnp.issubdtype(otype, dpnp.inexact):
+        pass
+    else:
+        # All other types convert to floating point.
+        # First check if f is a dpnp integer type; if so, convert f to default
+        # float type to avoid modular arithmetic when computing changes in f.
+        if dpnp.issubdtype(otype, dpnp.integer):
+            f = f.astype(dpnp.default_float_type())
+        otype = dpnp.default_float_type()
+
+    for axis_, ax_dx in zip(axes, dx):
+        if f.shape[axis_] < edge_order + 1:
+            raise ValueError(
+                "Shape of array too small to calculate a numerical gradient, "
+                "at least (edge_order + 1) elements are required."
+            )
+
+        # result allocation
+        if dpnp.isscalar(ax_dx):
+            usm_type = f.usm_type
         else:
-            if len(varargs) == 0:
-                return dpnp_gradient(x1_desc).get_pyobj()
+            usm_type = dpu.get_coerced_usm_type([f.usm_type, ax_dx.usm_type])
+        out = dpnp.empty_like(f, dtype=otype, usm_type=usm_type)
```

Comment:
Do not we need to check that `f` and `ax_dx` are on the same SYCL queue?
```
import  dpnp
dx=dpnp.arange(6.0)
x = dpnp.array([0, 2.0, 3.0, 4.0, 5.0, 5.0], dtype=dpnp.float32, device="cpu")
x = dpnp.tile(x, (6, 1)) + x.reshape(-1, 1)

dpnp.gradient(x, dx, axis=0)

File
~/repos/dpnp_review/dpnp/dpnp_iface_mathematical.py:235, in _gradient_num_diff_2nd_order_interior(f, ax_dx, out, slices, axis, uniform_spacing)
    232 slice4[axis] = slice(2, None)
    234 if uniform_spacing:
--> 235     out[tuple(slice1)] = (f[tuple(slice4)] - f[tuple(slice2)]) / (
    236         2.0 * ax_dx
    237     )
    238 else:
    239     dx1 = ax_dx[0:-1]
    
# ExecutionPlacementError: Execution placement can not be unambiguously inferred from input arguments.
```

Reply:
I guess no, because it's already done in elementwise functions, like in your example. And we would probably fine to detect `ExecutionPlacementError` at that point, during any elementwise operation.
Thus I believe we can avoid explicit check of allocation on the same queue for the input arrays inside `dpnp.gradient` call.